### ISSUE TITLE
@3.2.0 Expected error but got string at addRequestedAndFinish

### ISSUE THREAD

**Issue by mxdubois** [npm/npm#9206]
Hey, I'm getting an error when I run `npm link` with `npm@3.2.0`.

npm-debug.log
```
0 info it worked if it ends with ok
1 verbose cli [ 'node',
1 verbose cli   '/Users/expeditefinancial/dev/env-0.12.7/bin/npm',
1 verbose cli   'link' ]
2 info using npm@3.2.0
3 info using node@v0.12.7
4 verbose linkPkg /Users/expeditefinancial/dev/ls-portal/middleware-analytics
5 silly gentlyRm /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/middleware-analytics is being purged
6 verbose gentlyRm don't care about contents; nuking /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/middleware-analytics
7 silly vacuum-fs purging /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/middleware-analytics
8 silly vacuum-fs quitting because other entries in /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules
9 verbose link build target /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/middleware-analytics
10 silly loadCurrentTree Starting
11 silly install loadCurrentTree
12 silly install readLocalPackageData
13 silly install normalizeTree
14 silly loadCurrentTree Finishing
15 silly loadIdealTree Starting
16 silly install loadIdealTree
17 silly cloneCurrentTree Starting
18 silly install cloneCurrentTreeToIdealTree
19 silly cloneCurrentTree Finishing
20 silly loadShrinkwrap Starting
21 silly install loadShrinkwrap
22 silly loadShrinkwrap Finishing
23 silly loadAllDepsIntoIdealTree Starting
24 silly install loadAllDepsIntoIdealTree
25 silly fetchNamedPackageData redux-reducer-util
26 silly mapToRegistry name redux-reducer-util
27 silly mapToRegistry using default registry
28 silly mapToRegistry registry https://registry.npmjs.org/
29 silly mapToRegistry uri https://registry.npmjs.org/redux-reducer-util
30 verbose request uri https://registry.npmjs.org/redux-reducer-util
31 verbose request no auth needed
32 info attempt registry request try #1 at 9:50:58 PM
33 verbose request id e5689b32173f60a8
34 http request GET https://registry.npmjs.org/redux-reducer-util
35 http 404 https://registry.npmjs.org/redux-reducer-util
36 verbose headers { 'cache-control': 'max-age=0',
36 verbose headers   'transfer-encoding': 'chunked',
36 verbose headers   'accept-ranges': 'bytes',
36 verbose headers   date: 'Sun, 09 Aug 2015 04:50:07 GMT',
36 verbose headers   via: '1.1 varnish',
36 verbose headers   age: '0',
36 verbose headers   connection: 'keep-alive',
36 verbose headers   'x-served-by': 'cache-sjc3130-SJC',
36 verbose headers   'x-cache': 'MISS',
36 verbose headers   'x-cache-hits': '0',
36 verbose headers   'x-timer': 'S1439095807.707805,VS0,VE87' }
37 silly get cb [ 404,
37 silly get   { 'cache-control': 'max-age=0',
37 silly get     'transfer-encoding': 'chunked',
37 silly get     'accept-ranges': 'bytes',
37 silly get     date: 'Sun, 09 Aug 2015 04:50:07 GMT',
37 silly get     via: '1.1 varnish',
37 silly get     age: '0',
37 silly get     connection: 'keep-alive',
37 silly get     'x-served-by': 'cache-sjc3130-SJC',
37 silly get     'x-cache': 'MISS',
37 silly get     'x-cache-hits': '0',
37 silly get     'x-timer': 'S1439095807.707805,VS0,VE87' } ]
38 verbose stack Error: Argument #1: Expected error but got string
38 verbose stack     at addRequestedAndFinish (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/fetch-package-metadata.js:63:5)
38 verbose stack     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/iferr/index.js:11:16
38 verbose stack     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/utils/pulse-till-done.js:19:8
38 verbose stack     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/cache/caching-client.js:160:20
38 verbose stack     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/cache/caching-client.js:53:8
38 verbose stack     at f (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/once/once.js:17:25)
38 verbose stack     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:87:10
38 verbose stack     at f (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/once/once.js:17:25)
38 verbose stack     at CachingRegistryClient.<anonymous> (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:211:14)
38 verbose stack     at Request._callback (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:170:14)
39 verbose cwd /Users/expeditefinancial/dev/ls-portal/middleware-analytics
40 error Darwin 14.3.0
41 error argv "node" "/Users/expeditefinancial/dev/env-0.12.7/bin/npm" "link"
42 error node v0.12.7
43 error npm  v3.2.0
44 error code EINVALIDTYPE
45 error typeerror Error: Argument #1: Expected error but got string
45 error typeerror     at addRequestedAndFinish (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/fetch-package-metadata.js:63:5)
45 error typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/iferr/index.js:11:16
45 error typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/utils/pulse-till-done.js:19:8
45 error typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/cache/caching-client.js:160:20
45 error typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/cache/caching-client.js:53:8
45 error typeerror     at f (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/once/once.js:17:25)
45 error typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:87:10
45 error typeerror     at f (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/once/once.js:17:25)
45 error typeerror     at CachingRegistryClient.<anonymous> (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:211:14)
45 error typeerror     at Request._callback (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:170:14)
45 error typeerror This is an error with npm itself. Please report this error at:
45 error typeerror     <http://github.com/npm/npm/issues>
46 verbose exit [ 1, true ]
```

console output:
```
npm ERR! Darwin 14.3.0
npm ERR! argv "node" "/Users/expeditefinancial/dev/env-0.12.7/bin/npm" "link"
npm ERR! node v0.12.7
npm ERR! npm  v3.2.0
npm ERR! code EINVALIDTYPE

npm ERR! typeerror Error: Argument #1: Expected error but got string
npm ERR! typeerror     at addRequestedAndFinish (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/fetch-package-metadata.js:63:5)
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/iferr/index.js:11:16
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/utils/pulse-till-done.js:19:8
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/cache/caching-client.js:160:20
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/cache/caching-client.js:53:8
npm ERR! typeerror     at f (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/once/once.js:17:25)
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:87:10
npm ERR! typeerror     at f (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/once/once.js:17:25)
npm ERR! typeerror     at CachingRegistryClient.<anonymous> (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:211:14)
npm ERR! typeerror     at Request._callback (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:170:14)
npm ERR! typeerror This is an error with npm itself. Please report this error at:
npm ERR! typeerror     <http://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/expeditefinancial/dev/ls-portal/middleware-analytics/npm-debug.log
```

**Comment by mxdubois** [comment]
Upgraded to `npm@3.2.2` and got the same error from a different line:
```
npm ERR! Darwin 14.3.0
npm ERR! argv "node" "/Users/expeditefinancial/dev/env-0.12.7/bin/npm" "link"
npm ERR! node v0.12.7
npm ERR! npm  v3.2.2
npm ERR! code EINVALIDTYPE

npm ERR! typeerror Error: Argument #1: Expected error but got string
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/install/deps.js:227:5
npm ERR! typeerror     at cb (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/slide/lib/async-map.js:47:24)
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/install/deps.js:287:14
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/install/and-add-parent-to-errors.js:11:5
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/iferr/index.js:11:16
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/fetch-package-metadata.js:35:12
npm ERR! typeerror     at addRequestedAndFinish (/Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/fetch-package-metadata.js:69:5)
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/node_modules/iferr/index.js:11:16
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/utils/pulse-till-done.js:19:8
npm ERR! typeerror     at /Users/expeditefinancial/dev/env-0.12.7/lib/node_modules/npm/lib/cache/caching-client.js:160:20
npm ERR! typeerror This is an error with npm itself. Please report this error at:
npm ERR! typeerror     <http://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/expeditefinancial/dev/ls-portal/middleware-analytics/npm-debug.log
```

**Comment by KenanY** [comment]
@mxdubois Do you have `redux-reducer-util` as a dependency anywhere? Such a module does not exist in the registry:

```
34 http request GET https://registry.npmjs.org/redux-reducer-util
35 http 404 https://registry.npmjs.org/redux-reducer-util
```

**Comment by mxdubois** [comment]
Yeah. It's linked into my local registry.

But I've discovered the problem! I had a few dependency keys like this:
`"redux-reducer-util": "link"` -- which is invalid but used to work for me
in npm2.

Obviously I don't expect you to support my invalid keys but it might be
worth it to improve the error handling in this case. Up to you guys.

Thanks for the help!

**Comment by iarna** [comment]
So it turns out this bug was triggered by an update to `npm-registry-couchapp`– it'll probably yet get patched, but in the meantime, as of npm@3.2.2, npm will handle this gracefully.

Since you've apparently found the thing triggering the error for yourself as well, I'm gonna close this ticket. Thanks for taking the time to submit this!
